### PR TITLE
Add particle cloud display plugin for RViz

### PR DESCRIPTION
**Pull request by naiveHobo**
Signed-off-by: Sarthak Mittal <sarthakmittal2608@gmail.com>

---

## Basic Info

| Info | Please fill out this column |
| ------ | ----------- |
| Ticket(s) this addresses   | #1678 |
| Primary OS tested on | Ubuntu 18.04 |
| Robotic platform tested on | TurtleBot3 gazebo simulation |

---

## Description of contribution in a few bullet points

* Added Particle Cloud Display plugin for RViz to display AMCL particle clouds
* Particles scale according to weight
* Update bringup rviz files

---

## Future work that may be required in bullet points

* Move to the default rviz display plugin for nav2_msgs/ParticleCloud after support is available (ros2/rviz#538)

### CODE REVIEW

File: nav2_bringup/bringup/rviz/nav2_namespaced_view.rviz [hunk 1]
```diff
@@ -164,24 +164,19 @@ Visualization Manager:
...
-      Shaft Length: 0.23000000417232513
-      Shaft Radius: 0.009999999776482582
       Shape: Arrow (Flat)
       Topic:
         Depth: 5
         Durability Policy: Volatile
         History Policy: Keep Last
         Reliability Policy: Best Effort
-        Value: <robot_namespace>/particlecloud
+        Value: <robot_namespace>/particle_cloud
```
Comment (SteveMacenski) [inline nav2_bringup/bringup/rviz/nav2_namespaced_view.rviz:179]: I think this is the only change you actually need to make, can you revert the other rviz parameters? Or is there some reason the other updates are required?

File: nav2_rviz_plugins/CMakeLists.txt [hunk 2]
```diff
@@ -37,6 +37,8 @@ set(nav2_rviz_plugins_headers_to_moc
   include/nav2_rviz_plugins/goal_common
   include/nav2_rviz_plugins/goal_tool.hpp
   include/nav2_rviz_plugins/nav2_panel.hpp
+  include/nav2_rviz_plugins/flat_weighted_arrows_array.hpp
```
Comment (SteveMacenski) [inline nav2_rviz_plugins/CMakeLists.txt:40]: Can you put this in a separate directory `particle_filter_plugin` or something to keep them organized?

File: nav2_rviz_plugins/include/nav2_rviz_plugins/flat_weighted_arrows_array.hpp [hunk 3]
```diff
@@ -0,0 +1,95 @@
+/*
+ * Copyright (c) 2012, Willow Garage, Inc.
+ * Copyright (c) 2018, Bosch Software Innovations GmbH.
+ * Copyright (c) 2020, Sarthak Mittal
```
Comment (SteveMacenski) [inline nav2_rviz_plugins/include/nav2_rviz_plugins/flat_weighted_arrows_array.hpp:4]: Remove your name in the BSD license, you can only apply a license once and below is Apache 2.

File: nav2_rviz_plugins/include/nav2_rviz_plugins/particle_cloud_display.hpp [hunk 4]
```diff
@@ -0,0 +1,162 @@
+/*
+ * Copyright (c) 2012, Willow Garage, Inc.
+ * Copyright (c) 2018, Bosch Software Innovations GmbH.
+ * Copyright (c) 2020, Sarthak Mittal
```
Comment (SteveMacenski) [inline nav2_rviz_plugins/include/nav2_rviz_plugins/particle_cloud_display.hpp:4]: Same here, remove your name here, you have it included below in Apache

File: nav2_rviz_plugins/include/nav2_rviz_plugins/particle_cloud_display.hpp [hunk 5]
```diff
@@ -0,0 +1,162 @@
...
+#include <memory>
+#include <vector>
+
+#include "nav2_msgs/msg/particle_cloud.hpp"
+
+#include "rviz_rendering/objects/shape.hpp"
+#include "rviz_common/message_filter_display.hpp"
+
+// TODO(botteroa): Originally the display extended the MessageFilterDisplay. Revisit when available.
+// #include "rviz_common/message_filter_display.hpp"
```
Comment (SteveMacenski) [inline nav2_rviz_plugins/include/nav2_rviz_plugins/particle_cloud_display.hpp:59]: Just curious, is this ported yet? I think it was

File: nav2_rviz_plugins/src/particle_cloud_display.cpp [hunk 6]
```diff
@@ -0,0 +1,424 @@
+/*
+ * Copyright (c) 2012, Willow Garage, Inc.
+ * Copyright (c) 2018, Bosch Software Innovations GmbH.
+ * Copyright (c) 2020, Sarthak Mittal
```
Comment (SteveMacenski) [inline nav2_rviz_plugins/src/particle_cloud_display.cpp:4]: name

File: nav2_bringup/bringup/rviz/nav2_namespaced_view.rviz (hunk 1 above)
Comment (naiveHobo) [reply nav2_bringup/bringup/rviz/nav2_namespaced_view.rviz:179]: This plugin scales the length and radius of the arrows according to the weight of the particle. What that means is that each particle is represented with an arrow of possibly a different length based on its weight. Those parameters that I have removed defined the length and radius of all the arrows (like how it is with PoseArray display).

I have instead defined two parameters, minimum and maximum arrow length. The arrows are scaled to fit this range based on their weights.

File: nav2_rviz_plugins/include/nav2_rviz_plugins/flat_weighted_arrows_array.hpp (hunk 3 above)
Comment (naiveHobo) [reply nav2_rviz_plugins/include/nav2_rviz_plugins/flat_weighted_arrows_array.hpp:4]: Ah, sorry about this! I'd removed it from one file and forgot to remove it from the other files.

File: nav2_bringup/bringup/rviz/nav2_namespaced_view.rviz (hunk 1 above)
Comment (SteveMacenski) [reply nav2_bringup/bringup/rviz/nav2_namespaced_view.rviz:179]: Why did that involve resizing the window, disabling the TF tree, state, update topic, etc. I'm not concerned about the params under `nav2_rviz_plugins/ParticleCloudDisplay`, but the other things you modified in the rviz configs if you look at the diff

File: nav2_rviz_plugins/include/nav2_rviz_plugins/particle_cloud_display.hpp (hunk 5 above)
Comment (naiveHobo) [reply nav2_rviz_plugins/include/nav2_rviz_plugins/particle_cloud_display.hpp]: Yup, looks like it has been ported. Removed the TODO.

**Comment by naiveHobo** [comment]
p1

p2

p3

p4

File: nav2_rviz_plugins/include/nav2_rviz_plugins/particle_cloud_display.hpp (hunk 5 above)
Comment (SteveMacenski) [reply nav2_rviz_plugins/include/nav2_rviz_plugins/particle_cloud_display.hpp]: did you reinclude it / use it?

**Comment by SteveMacenski** [comment]
@naiveHobo I don't notice any difference between this and the PoseArray visualization. I'm not saying that's a problem, but if you wanted to highlight the changes in weights, you may want to consider making it more clear. 

Let me know. Otherwise, I'm good with this. You should open a PR on RVIZ2 to add this.

**Comment by SteveMacenski** [comment]
Great job on this and fast turn around :1st_place_medal:

**Comment by naiveHobo** [comment]
> @naiveHobo I don't notice any difference between this and the PoseArray visualization. I'm not saying that's a problem, but if you wanted to highlight the changes in weights, you may want to consider making it more clear.

I was testing why this looks so similar to the PoseArray visualization and the scaling is not apparent based on the weights. Seems to me that this might be a bug in AMCL. The weights are always evenly distributed among all the particles. When I say evenly distributed, it means they are exactly the same for every particle in the cloud. Hence, the arrows all look the same in the visualization.

@SteveMacenski as far as I can tell, that's not the expected behavior, right?

> Let me know. Otherwise, I'm good with this. You should open a PR on RVIZ2 to add this.

I was about to open a PR on rviz2, but I'd have to add nav2_msgs as a dependency for rviz_default_plugins in rviz2. If I'm understanding correctly, that's not possible right not, right? The default ros2 packages should build without any nav2 dependencies right now, right?


> Great job on this and fast turn around 1st_place_medal Let me know if you're interested in doing something next or what's on your mind. I recall your interest in control and AI. There's also a body of software work as well if you're interested in continuing in this vein. Obviously, up to you +1 Your help is appreciated.

Thanks, love to be of help! I'd love to contribute to anything that needs help really.

File: nav2_rviz_plugins/include/nav2_rviz_plugins/particle_cloud_display.hpp (hunk 5 above)
Comment (naiveHobo) [reply nav2_rviz_plugins/include/nav2_rviz_plugins/particle_cloud_display.hpp]: It looks like this todo has mistakenly not been removed from rviz_default_plugins package itself.
The message filter display include was there already.

**Comment by SteveMacenski** [comment]
> that's not the expected behavior

Yeah, I'd say so. 

> nav2_msgs as a dependency for rviz_default_plugins in rviz2

True, I'd be interested in their thoughts on it. We can spin it out into another repo or something, I don't mind, but I think they'll have something they want in mind so rather than guessing, lets just find out. Its also already dependent on `nav_msgs` so I can make a really good argument that this should be added too. If they don't like that, we can literally just put the particle messages _into_ nav_msgs and no new dependencies. 

> Thanks, love to be of help! I'd love to contribute to anything that needs help really.

Awesome, let me think about this for a bit. Anything you want to learn about in ROS2 / navigation / software? Just to help narrow the field. If you're at all interested in the launch system (python) and lifecycle, there's a few things there that could be good. Working on some AI demos could be really good reference designs. Leading the ship on a new controller project could be a good opportunity if you had ideas / thoughts on that. CiLQR, iLQR, MPC, etc keep coming up. I'm working (well, I _was_ before my days got consumed with replying to tickets / PRs / slack) on a motion planner that touches non-linear optimization, search, and software design. All are options. Also anything else you see on GItHub or on the pinned posts on Slack channels.

File: nav2_rviz_plugins/include/nav2_rviz_plugins/particle_cloud_display.hpp (hunk 5 above)
Comment (SteveMacenski) [reply nav2_rviz_plugins/include/nav2_rviz_plugins/particle_cloud_display.hpp]: ah ok

**Comment by SteveMacenski** [comment]
Ah, codecov to the reminder can you write some tests for this? I have to be honest, I have no clue how to write rviz display tests....

.... luckily for us this resembles closely pose array and they have tests for that https://github.com/ros2/rviz/tree/ros2/rviz_default_plugins/test/rviz_default_plugins/displays/pose_array to steal + convert

**Comment by naiveHobo** [comment]
Turns out rviz has a brilliant library for visual testing. I had to port over a few files as they were from rviz_default_plugins, though.

> Leading the ship on a new controller project could be a good opportunity if you had ideas / thoughts on that. CiLQR, iLQR, MPC, etc keep coming up. I'm working (well, I _was_ before my days got consumed with replying to tickets / PRs / slack) on a motion planner that touches non-linear optimization, search, and software design.

I'd love to help with the controller development. Your project sounds extremely interesting, would love to be a part there. I have worked with local planners/controllers before (pure pursuit and DWA) so I think the learning curve wouldn't be too steep for me and I can get to development quickly.

**Comment by SteveMacenski** [comment]
Got it, hopefully they don't live here too long and we can just merge it into rviz2 right away. It might be worth putting a PR into Rviz2 now that you have some tests as well (once they work). 

Your build seems to be failing on a linking issue. Might need to link against some rviz libraries. 

Lets chat in slack 

- S

**Comment by naiveHobo** [comment]
Can't seem to figure out the linking issue. There are undefined references to ```rviz_default_plugins::tools::PoseTool``` but PoseTool is only used with ```nav2_rviz_plugins::GoalTool``` and there haven't been any library linking changes in nav2_rviz_plugins. Plus, I can't reproduce the linking issue on my system.

@SteveMacenski @shivaang12 any idea what this could be about?

**Comment by SteveMacenski** [comment]
It may be because you didn't make the 2 rviz plugins different libraries. It looks like you appended your file to the same library as the pose tool and panel. Try keeping different libraries. 
In the test CMake section you find packages like `rviz_default_plugins` but you don't link against them.

**Comment by SteveMacenski** [comment]
That seems to have fixed it, what ended up being the issue?

**Comment by naiveHobo** [comment]
I think it's pretty much what you suggested. Since I had added the display plugin to the same library as panel and goal tool, during the test build I'm linking the complete library to the tests. During the tests, it's not able to find the links for panel and goal tool since I'd written the CMakeLists.txt assuming I'd only have to find the packages that the display plugin needed during the tests.

**Comment by SteveMacenski** [comment]
Well, that's what we have CI for :+1:

**Comment by naiveHobo** [comment]
I'm not sure what the codecov errors are. There's a small problem with plugin loading, so I'd hold the merge till the next commit.

**Comment by SteveMacenski** [comment]
Codecov error just means that you didnt make the test coverage higher. That was happening because of I think your loading issue, it didnt show any coverage for your additions.

**Comment by naiveHobo** [comment]
Ah, I just realised the test coverage is an issue. The rviz plugin tests are based on the availability of a display. The tests only run if a display is available, as per these lines.
```
# Tests currently only run on OS X @ OSRF jenkins
# Enable on Linux by providing a displays, enable on Windows via EnableDisplayTests=True
option(EnableDisplayTests "EnableDisplayTests")
set(DisplayTests "False" CACHE STRING "DisplayTestsVariable")

if(DEFINED ENV{DISPLAY})
  set(DISPLAYPRESENT TRUE)
endif()

if(APPLE OR DISPLAYPRESENT OR EnableDisplayTests STREQUAL "True")
  message(STATUS "Enabling tests requiring a display")
else()
  set(SKIP_DISPLAY_TESTS "SKIP_TEST")
endif()

# options and directories for visual tests (see visual_testing_framework documentation)
option(EnableVisualTests "decides whether or not to enable the tests")
```

**Comment by naiveHobo** [comment]
> I also don't want you wasting too much time trying to get these tests working in Nav2 since I really want this to live in Rviz2. If you can put stuff there and have it work, lets just do that and move on.

Makes sense. I shall revert back to the commits before I added the tests and move on I guess?

**Comment by SteveMacenski** [comment]
Sounds good. Please do submit that PR to RVIZ2. I dont think it will be merged before Foxy since they're all busy, but it gives them time to review it. RVIZ2 PRs typically take a while to merge so that's why I want you to submit it sooner than later so we can get that in. And potentially if we can get in before foxy, we just don't include it in this repo at all.

**Comment by naiveHobo** [comment]
I have submitted the PR (https://github.com/ros2/rviz/pull/541). Everything should be good to go after the revert.

**Comment by SteveMacenski** [comment]
Do the revert and lets wait a few days one way or another

**Comment by SteveMacenski** [comment]
Why close this?

**Comment by naiveHobo** [comment]
Ah, sorry closed this by mistake.

**Comment by SteveMacenski** [comment]
I'm of a mind that we just merge this. Can you pull in master so we can trigger a modern CI build? If it passes, lets just merge it. 

We've waited more than enough time to get the nav msgs stuff done if it were ever going to be. We should move on

**Comment by naiveHobo** [comment]
@SteveMacenski good to go?

**Comment by SteveMacenski** [comment]
Yup! Thanks for baring with me on this

**Pull request merged**